### PR TITLE
move crdmanager to its own module

### PR DESCRIPTION
**Pull request by reshnm**
**How to categorize this PR?**

/area usability
/kind enhancement
/priority 3

**What this PR does / why we need it**:
This PR moves the CRD manager to its own go module.
With this change, the CRD manager can be re-used in other (external) modules like the LAAS.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Review by schrodit — CHANGES_REQUESTED** [review]
For the submodule to work correctly you also have to adjust some more files:
1. The pipelines file to generate another tag on release https://github.com/gardener/landscaper/blob/master/.ci/pipeline_definitions#L93 (this is needed for go submodules)
2. The go.mod altering files for releases and the next version (this is needed so that consuming modules do not ned to manually set a go.mod overwrite(`require`)).
  https://github.com/gardener/landscaper/blob/master/.ci/next_version#L23
  https://github.com/gardener/landscaper/blob/master/.ci/release#L23

  Specifically these files need to be changed so that
  - the dependency towards the CRD manager in the landscaper's go.mod file is updated
  - the dependency towards the api in the CRD manager's go.mod file is updated

File: crdmanager/pkg/crdmanager/crdmanager.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	readerBufferSize = 32
```
[the remark this answers is not in the document]
Comment (reshnm) [reply crdmanager/pkg/crdmanager/crdmanager.go:32]: I think the name is a bit misleading. It is more of a parser read ahead size to determine wether the crd file is of type yaml or json.
I will rename it and increas its size to 1024. That should be safe.

File: crdmanager/pkg/crdmanager/crdmanager.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (crdmgr *CRDManager) createCrd(ctx context.Context, crd *v1.CustomResourceDefinition) error {
+	return crdmgr.client.Create(ctx, crd)
+}
```
[the remark this answers is not in the document]
Comment (reshnm) [reply crdmanager/pkg/crdmanager/crdmanager.go:138]: I have inlined the function.

**Comment by reshnm** [comment]
> For the submodule to work correctly you also have to adjust some more files:
> 
>     1. The pipelines file to generate another tag on release https://github.com/gardener/landscaper/blob/master/.ci/pipeline_definitions#L93 (this is needed for go submodules)
> 
>     2. The go.mod altering files for releases and the next version (this is needed so that consuming modules do not ned to manually set a go.mod overwrite(`require`)).
>        https://github.com/gardener/landscaper/blob/master/.ci/next_version#L23
>        https://github.com/gardener/landscaper/blob/master/.ci/release#L23
> 
> 
> Specifically these files need to be changed so that
> 
>     * the dependency towards the CRD manager in the landscaper's go.mod file is updated
> 
>     * the dependency towards the api in the CRD manager's go.mod file is updated

Thanks for pointing that out.
We just need to update the landscaper go.mod file, since the crdmanager is not used by the apis module.
Does it look good now?

File: crdmanager/pkg/crdmanager/crdmanager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (reshnm) [reply crdmanager/pkg/crdmanager/crdmanager.go]: I decided to stick with the name `readerBufferSize` as the parser API uses the same name.

**Comment by schrodit** [comment]
> > For the submodule to work correctly you also have to adjust some more files:
> > ```
> > 1. The pipelines file to generate another tag on release https://github.com/gardener/landscaper/blob/master/.ci/pipeline_definitions#L93 (this is needed for go submodules)
> > 
> > 2. The go.mod altering files for releases and the next version (this is needed so that consuming modules do not ned to manually set a go.mod overwrite(`require`)).
> >    https://github.com/gardener/landscaper/blob/master/.ci/next_version#L23
> >    https://github.com/gardener/landscaper/blob/master/.ci/release#L23
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Specifically these files need to be changed so that
> > ```
> > * the dependency towards the CRD manager in the landscaper's go.mod file is updated
> > 
> > * the dependency towards the api in the CRD manager's go.mod file is updated
> > ```
> 
> Thanks for pointing that out. We just need to update the landscaper go.mod file, since the crdmanager is not used by the apis module. Does it look good now?

I meant that you also have to adopt the crd manager's go.mod file like the following.
This is because it also used the landscaper api module just like the landscaper itself.
```
sed -i -e "0,/)/{s@github.com/gardener/landscaper/apis .*@github.com/gardener/landscaper/apis v0.0.0-00010101000000-000000000000@}" \
  ${SOURCE_DIR}/crdmanager/go.mod
```

**Comment by reshnm** [comment]
> > > For the submodule to work correctly you also have to adjust some more files:
> > > ```
> > > 1. The pipelines file to generate another tag on release https://github.com/gardener/landscaper/blob/master/.ci/pipeline_definitions#L93 (this is needed for go submodules)
> > > 
> > > 2. The go.mod altering files for releases and the next version (this is needed so that consuming modules do not ned to manually set a go.mod overwrite(`require`)).
> > >    https://github.com/gardener/landscaper/blob/master/.ci/next_version#L23
> > >    https://github.com/gardener/landscaper/blob/master/.ci/release#L23
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Specifically these files need to be changed so that
> > > ```
> > > * the dependency towards the CRD manager in the landscaper's go.mod file is updated
> > > 
> > > * the dependency towards the api in the CRD manager's go.mod file is updated
> > > ```
> > 
> > 
> > Thanks for pointing that out. We just need to update the landscaper go.mod file, since the crdmanager is not used by the apis module. Does it look good now?
> 
> I meant that you also have to adopt the crd manager's go.mod file like the following. This is because it also used the landscaper api module just like the landscaper itself.
> 
> ```
> sed -i -e "0,/)/{s@github.com/gardener/landscaper/apis .*@github.com/gardener/landscaper/apis v0.0.0-00010101000000-000000000000@}" \
>   ${SOURCE_DIR}/crdmanager/go.mod
> ```

got it

**Review by schrodit — APPROVED** [review]
looks good from my side

**Pull request merged**